SimExLite/PropagationCalculators/WPGProgagationCalculator.py

```diff
@@ -0,0 +1,112 @@
+""":module WPGPropagationCalculator: Module that holds the WPGPropagationCalculator class."""
+
+from pathlib import Path
+import shutil
+import sys
+
+# Requires WPG package: https://github.com/samoylv/WPG
+from s2e.prop import propagate_s2e
+
+from libpyvinyl import BaseCalculator, CalculatorParameters
+from libpyvinyl.BaseData import DataCollection
+from SimExLite.WavefrontData import WavefrontData, WPGFormat
+
+
+class WPGPropagationCalculator(BaseCalculator):
+    """:class WPGPropagationCalculator: Class representing photon propagation through X-ray optics."""
+
+    def __init__(
+        self,
+        name: str,
+        input: DataCollection,
+        output_keys: str = "WPG_wavefront",
+        output_data_types=WavefrontData,
+        output_filenames: str = "wavefront.h5",
+        instrument_base_dir="./",
+        calculator_base_dir="WPGPropagationCalculator",
+        parameters=None,
+    ):
+        super().__init__(
+            name,
+            input,
+            output_keys,
+            output_data_types=output_data_types,
+            output_filenames=output_filenames,
+            instrument_base_dir=instrument_base_dir,
+            calculator_base_dir=calculator_base_dir,
+            parameters=parameters,
+        )
+        self.__WPG_path = None
+
+    def init_parameters(self):
+        parameters = CalculatorParameters()
+        beamline_config = parameters.new_parameter(
+            "beamline_config_file", comment="The beamline_configfile"
+        )
+        Path(self.base_dir).mkdir(parents=True, exist_ok=True)
+        simple_beamline_fn = str(Path(self.base_dir) / "simple_beamline.py")
+        beamline_config.value = create_simple_beamline_file(simple_beamline_fn)
+
+        self.parameters = parameters
+
+    def prep_beamline_config(self):
+        """Copy the beamline config file to the working dir to import the beamline module"""
+        beamline_config_fn = self.parameters["beamline_config_file"].value
+        dst_path = Path(self.base_dir) / "WPG_beamline.py"
+        shutil.copyfile(beamline_config_fn, str(dst_path))
+
+    def get_input_fn(self):
+        """Make sure the data is a mapping of WPGFormat file"""
+        assert len(self.input) == 1
+        input_data = self.input.to_list()[0]
+        if input_data.mapping_type == WPGFormat:
+            input_fn = input_data.filename
+        else:
+            filepath = Path(self.base_dir) / "input_wavefront.h5"
+            input_data.write(str(filepath), WPGFormat)
+            input_fn = str(filepath)
+
+        return input_fn
+
+    def backengine(self):
+
+        self.prep_beamline_config()
+        sys.path.insert(0, self.base_dir)
+        import WPG_beamline
+
+        input_fn = self.get_input_fn()
+        output_fn = str(Path(self.base_dir) / self.output_filenames[0])
+
+        propagate_s2e.propagate(input_fn, output_fn, WPG_beamline.get_beamline)
+
+        assert len(self.output_keys) == 1
```

Comment:
This would raise an exception if condition is False. Should return error message and abort the calculation more gently.

Reply:
`assert` is deprecated